app/screens/browse_channels/browse_channels.tsx

```diff
@@ -0,0 +1,501 @@
+    const theme = useTheme();
+    const style = getStyleFromTheme(theme);
+    const serverUrl = useServerUrl();
+
+    const [channels, setChannels] = useState<Channel[]>([]);
+    const [archivedChannels, setArchivedChannels] = useState<Channel[]>([]);
+    const [sharedChannels, setSharedChannels] = useState<Channel[]>([]);
+
+    const [visibleChannels, setVisibleChannels] = useState<Channel[]>([]);
+    const [typeOfChannels, setTypeOfChannels] = useState(PUBLIC);
+
+    const [term, setTerm] = useState('');
+
+    const [adding, setAdding] = useState(false);
+    const [loading, setLoading] = useState(false);
+
+    const mounted = useRef(false);
+    const publicPage = useRef(-1);
+    const sharedPage = useRef(-1);
+    const archivedPage = useRef(-1);
+    const nextPublic = useRef(true);
+    const nextShared = useRef(true);
+    const nextArchived = useRef(true);
+
+    const setHeaderButtons = (createEnabled: boolean) => {
+        const buttons = {
+            leftButtons: [makeLeftButton(closeButton)],
+            rightButtons: [] as OptionsTopBarButton[],
+        };
+
+        if (canCreateChannels) {
+            buttons.rightButtons = [{...makeRightButton(theme, intl.formatMessage), enabled: createEnabled}];
+        }
+
+        setButtons(componentId, buttons);
+    };
+
+    const onSelectChannel = async (channel: Channel) => {
+        setHeaderButtons(false);
+        setAdding(true);
+
+        const result = await joinChannel(serverUrl, currentUserId, currentTeamId, channel.id, '', false);
+
+        if (result.error) {
+            alertErrorWithFallback(
+                intl,
+                result.error,
+                {
+                    id: 'mobile.join_channel.error',
+                    defaultMessage: "We couldn't join the channel {displayName}. Please check your connection and try again.",
+                },
+                {
+                    displayName: channel ? channel.display_name : '',
+                },
+            );
+            setHeaderButtons(true);
+            setAdding(false);
+        } else {
+            close();
+            switchToChannel(serverUrl, channel.id, currentTeamId);
+        }
+    };
+
+    const doGetChannels = useCallback(() => {
+        if (!loading && mounted.current) {
+            const loadedChannels = async (data: Channel[] | undefined) => {
+                if (mounted.current) {
+                    switch (typeOfChannels) {
+                        case PUBLIC: {
+                            publicPage.current += 1;
+                            nextPublic.current = Boolean(data?.length);
+                            const filtered = filterJoinedChannels(joinedChannels, data);
+                            if (filtered?.length) {
+                                setChannels([...channels, ...filtered]);
+                            }
+                            if (data?.length && !filtered?.length) {
+                                doGetChannels();
```

Comment:
I should revisit this, since this is working for the wrong reasons.
The instance of `doGetChannels` we have here is the original one, which had `loading = false`.

Also, if we switch channel types while it is loading, it may break the flow too. For that, we may consider disabling changing the channels to show when it is loading.

Reply:
yeah disabling the switch should be fine as is loading